### ISSUE TITLE
Add Sidebar Management to site settings for multi-journal installs

### ISSUE THREAD

**Issue by asmecher** [pkp/pkp-lib#1859]
See http://forum.pkp.sfu.ca/t/ojs-3-0-choose-language-search-at-the-index-page/20864

**Comment by asmecher** [comment]
(Also the Footer field -- see http://forum.pkp.sfu.ca/t/configure-page-footer-in-multijournal-homepage/21496)

**Comment by NateWr** [comment]
Tests running.

This required that I dip into the authorization setup a little bit, so I'd appreciate a close look at this. Particularly [this](https://github.com/pkp/pkp-lib/pull/1914/commits/bfee4030a7404b3d55c4d4416f9381131701233a#diff-fb4ac974ab13c9d87aaa2a0dfa9766e4L25).

Also, the changes to [SiteSetupForm.inc.php](https://github.com/pkp/pkp-lib/pull/1914/commits/bfee4030a7404b3d55c4d4416f9381131701233a#diff-7716fdb0d5ad2c815c3741a2fe5e2be9R173) duplicate code in PKPAppearanceForm.inc.php.

PRs:
https://github.com/pkp/pkp-lib/pull/1914
https://github.com/pkp/ojs/pull/1073

**Comment by NateWr** [comment]
Tests passed!

**Comment by NateWr** [comment]
@bozana code review?

**Comment by bozana** [comment]
Yes, sure :-)

**Comment by bozana** [comment]
@NateWr, does this have to be migrated for OMP as well?

**Comment by bozana** [comment]
Ah, OMP has no site management... Sorry! :-(

**Comment by NateWr** [comment]
No worries. I think they're going to put site management in place for OMP at some point...

**Comment by bozana** [comment]
@NateWr, everything merged. The PKPSiteSettingsForm and SiteSetupForm should be cleaned up, but maybe to open a new issue for that? -- also depending on the decisions for OMP -- I think first there was such a feature for OMP, then it was removed, now maybe it will come again... :-)
And this issue can be closed, right?

**Comment by NateWr** [comment]
Maybe this should be cherry-picked... leave open for now.

**Comment by NateWr** [comment]
I was just updating with some changes for #1765 and I found a bug with this. If you edit the sidebar from the site settings panel, but you do it with a specific journal context loaded, it updates that journal context, not the overall site.

This needs further work.

**Comment by NateWr** [comment]
A fix for this is waiting on the Site Settings form to be fixed. https://github.com/pkp/pkp-lib/issues/1958#issuecomment-259700288

**Comment by asmecher** [comment]
(The hold-up on #1958 should be resolved.)

**Comment by NateWr** [comment]
Yep, that fixes it! :+1:

**Comment by NateWr** [comment]
I spent a while on this today and I'm not sure I've got the best solution. I ended up adjusting a series of methods for block plugins to circumvent the automated context discovery function `getSettingMainContext` here:

https://github.com/pkp/pkp-lib/blob/master/classes/plugins/BlockPlugin.inc.php#L51-L92

The main context has to be overridden on the site settings form because the admin pages can be loaded from within a specific journal context. It also means I have to use a new component just to load the correct data and permissions.

I wonder if it makes more sense to solve the problem by making sure that the admin pages never include a given context. This makes the most sense given the structure of our settings pages, but it would entail some more work to make sure that the journal settings wizards (which are inside admin) still work.

Still, here's some PRs to look at my initial approach:

https://github.com/pkp/pkp-lib/pull/1988
https://github.com/pkp/ojs/pull/1097

**Comment by NateWr** [comment]
@asmecher Above is the comment I mentioned about the site-wide sidebar blocks.

**Comment by asmecher** [comment]
Ah. Yes, there's some nasty business there that used to be necessary back when pkp-lib needed to support a variety of context "depths" for the sake of OHS (where all operations take place without a "context"), OJS (where you have a single level of context, the Journal) and OCS (where you have up to two levels of context, the Conference and the Scheduled Conference).

Even if we do decide to rewrite OCS using the current pkp-lib, we'll get rid of that double-level approach. So all of that complexity can go -- you can use a single context ID, which can be zero if you're looking at the side-wide context. Does that help? Would you like some help in cleaning all that cruft out?

**Comment by NateWr** [comment]
@asmecher I refactored the plugins to not use `contextDepth`, getting rid of `getSettingMainContext()` and some other stuff.

However, I did a search for `contextDepth` and found a lot of remnants of it deep in the system (Application, Router, etc) that I didn't want to touch. Not sure if you want to touch that or not but I don't think I should. :)

I've updated the PRs (tests running). Mind if I assign this one to you?

**Comment by asmecher** [comment]
Sure, assign away. (Are the PRs the same as linked above?) I can take care of cleaning out the other context depth stuff -- there's a lot of ancient history to it.

**Comment by asmecher** [comment]
(@ctgraham, we spoke a little about this during the TC call.)

**Comment by NateWr** [comment]
@asmecher Yeah, still the same PRs in https://github.com/pkp/pkp-lib/issues/1859#issuecomment-260411934

**Comment by asmecher** [comment]
Moving to 3.0.2 -- a little too disruptive for a late-stage merge into `master` for 3.0.1.

**Comment by NateWr** [comment]
(updated for a fix identified by the tests. tests running again.)

**Comment by NateWr** [comment]
@asmecher So the tests are failing on the static pages tests. I tried enabling the static pages plugin in my local install and it causes the Settings > Journal > Masthead tab to break. It loads nothing. I've checked my error log and as far as I can tell there's no fatal errors. Just the usual warnings and notices I get with every page load (dozens of them due to PHP 7 standards).

Can you reproduce this problem? It's happening to me on `master`.

**Comment by asmecher** [comment]
Have you got the most recent static pages plugin checked out? Might need to pull the master branch over there.

**Comment by NateWr** [comment]
I checked out the master branch of the static pages plugin. Still couldn't reproduce any problem there. @asmecher are you suggesting that I should include a submodule commit for the static pages plugin bringing it up to master?

I did find a syntax error introduced in 3d68ac9b717a25c2df9c9a6360988f53dca64e7e. For now I've fixed that and restarted the tests to see where we get.

The PRs are still the same, but I'll dump them here just for easier reference:
https://github.com/pkp/pkp-lib/pull/1988
https://github.com/pkp/ojs/pull/1097

**Comment by bozana** [comment]
new PRs:
pkp-lib master: https://github.com/pkp/pkp-lib/pull/2127
ojs master: https://github.com/pkp/ojs/pull/1161

**Comment by bozana** [comment]
merged, can this be closed, or shall this be cherry-picked?

**Comment by asmecher** [comment]
It looks like it at least needs porting over to OMP, no?

**Comment by bozana** [comment]
No, I first also thought so, but OMP has no site :-)

**Comment by NateWr** [comment]
Yep! The changes to LanguageBlockToggle should be compatible so that no changes are needed in OMP. And when OMP gets a site-wide context, we'll need to build that site settings panel up fresh anyway, so we can handle that then I think.

**Comment by asmecher** [comment]
Thanks, @bozana and @NateWr!

**Comment by NateWr** [comment]
The CustomBlockManager plugin needs some adjustments to work properly in the site-wide context. The following PRs address issues with creating new site-wide blocks and managing their enabled/disabled state.

PRs:
https://github.com/pkp/customBlockManager/pull/21
https://github.com/pkp/pkp-lib/pull/2456
https://github.com/pkp/ojs/pull/1359

Reported: http://forum.pkp.sfu.ca/t/custom-block-manager-and-static-pages-on-site-level-in-ojs-3-0-2/29908

@asmecher The tests don't seem to be running on the OJS PR. Maybe because I included multiple submodule updates? What's the correct approach to running tests against multiple submodule changes?

**Comment by bozana** [comment]
@NateWr, everything looks good to me :-) I only think that you will have to add the submodule references anew :-( Then you can merge... :-)

**Comment by ctgraham** [comment]
Multiple submodule commits in the last commit (or merge commit) will break the startSubmodulesTRAVIS script.
https://github.com/pkp/ojs/blob/f2426b95ffab88d86d0100d54541a76f4e334c62/tools/startSubmodulesTRAVIS.sh
Line 19 will pull all of the subproject commit hashes, then line 22 will fail because $strLength is not exactly 40.

**Comment by NateWr** [comment]
Thanks @ctgraham!

@bozana tests are now running. Fingers crossed. :)

**Comment by bozana** [comment]
merged :-) maybe then just to update the reference to the customBlockManager?

**Comment by NateWr** [comment]
Thanks! I think we've got the submodule updated. I just added it in a separate commit here:

https://github.com/pkp/ojs/commit/9b893efa165052249859521ed752159c40ef5920

**Comment by bozana** [comment]
oh yes, yes... :-) so the issue can be closed?
